### DISCUSSION TITLE
`useFieldArray` not accepting valid name

### DISCUSSION THREAD

**Discussion by RonnyLark** [react-hook-form/react-hook-form#9404] [Q&A]
I have the following example:
https://codesandbox.io/s/morning-monad-s1g732?file=/src/App.tsx

In this example, I want to make a component that takes a `control`, and always uses the same field name in `useFieldArray`.
For some reason, even though the interfaces `FormState` and `FormStateWithItems` match, RHF throws a type error on line 32:
```
Type 'Resolver<FormState, any>' is not assignable to type 'Resolver<FormStateWithItems, any>'.ts(2322)
```

If I let go of this idea, remove the `Omit` type, and just use the normal method of passing a name, The the `useFieldArray` throws a type error (see `SecondControlledComponent` in the sandbox):
```
Type 'string' is not assignable to type 'never'.ts(2322)
```

I can't win with this.
I don't understand how to use `useFieldArray` in a nested component.
The closest discussions I found had nothing to do with this issue, talking about how the field must be an array (which it is) and that the field name must be a const (which it doesn't, as seen in the example).

**Comment by tparvi** [comment]
Did you find a solution? I am running into same issue.

**Reply by RonnyLark** [reply]
I did not. I resorted to tricking RHF into accepting a hard-coded name by overriding the interface.
```ts
export type MultiInputField = { value: string }[];
type FormMock = { field: MultiInputField };
type ArrayPropsOverride = UseFieldArrayProps<FormMock, 'field'>;

useFieldArray({ control, name, rules } as unknown as ArrayPropsOverride)
```
This way I can have the component accept whatever name I want, while inside the component TS would still see the correct field type.

**Comment by martin-wahlberg** [comment]
I solved your issue by doing it this way:

This makes the ts compiler recognize that the only possible value for props is a controller with the required fields. Since we know that we are getting a controller for FormStateWithItems passed here the useFieldArrayHook will be satisfied with the name "items" since the corresponding field in the form is of type array.

```typescript react
interface Props<Form extends FormStateWithItems> {
  control: Form extends FormStateWithItems ? Control<Form> : never;
}

function ControlledComponent<Form extends FormStateWithItems>({
  control
}: Props<Form>) {
  useFieldArray({ control, name: "items" });
  return <div />;
}
```
In your second problem the issue was that you didn't ensure that the path (name) was for a field of an Array type. This is something that you have to ensure when using fieldArray.

```typescript react
interface SecondProps<F extends FieldValues, Name extends ArrayPath<F>> {
  control: Control<F>;
  name: Name;
}

function SecondControlledComponent<
  F extends FieldValues,
  Name extends ArrayPath<F>
>({ control, name }: SecondProps<F, Name>) {
  const fields = useFieldArray({ control, name });
  return <div />;
}
```

I updated your code sandbox here:
https://codesandbox.io/s/condescending-swanson-rswf9p?file=/src/App.tsx:376-496

I also have solved some similar issues here: 
https://github.com/react-hook-form/react-hook-form/discussions/8924 
This solution would probably work better for you as well since it helps you in avoiding getting values typed as any

Hope this might help you as well @tparvi
